### ISSUE TITLE
Exporting the gradient to MontePython

### ISSUE THREAD

**Issue by nnedelko** [Lbalkenhol/candl#1]
Hi! I'm working on modifying MontePython with an HMC/NUTS solver, which requires getting the gradient of loglkl. I would like to use candl for the CMB data, but I don't see a function that would directly give me the gradient vector. Is there a simple way to add this as a function in the likelihood classes (basically a log_like_grad(self, params))?

**Comment by Lbalkenhol** [comment]
Hello and thanks for your interest in using candl! Having access to gradient based samplers in MontePython would be amazing.

The simplest way to get gradients would be to do `jax.jacrev(like_func)` where `like_func` is a function that moves from parameters to logL values, i.e. first calls a differentiable theory code (e.g. CosmoPower emulators) and then hands the result off to candl in one step. I would warn against trying to do this piece-wise (theory and likelihood independently), as the input of cand likelihoods is relatively large (unbinned CMB spectra) and thus this calculation is computationally expensive.

**Comment by nnedelko** [comment]
Thank you for the answer! I am currently using a modified version of CLASS with improved non-linear contributions for my theory calculations for several extensions of LCDM. So if I understand the CosmoPower method correctly, the most effective way is to train my own emulators for each cosmological model using CLASS and then use these new emulators with candl, correct?

**Comment by nnedelko** [comment]
That does sound like the faster way, but I think for general accessibility being able to fall back to running directly with a Boltzmann solver is important. So will I still be able to use JAX on the result of `get_params_to_logl_func` if I use it with `get_CLASS_pars_to_theory_specs_func` rather than `get_CosmoPower_pars_to_theory_specs_func`?

**Comment by Lbalkenhol** [comment]
Yes that's the idea re CosmoPower. Though note that a suite of pre-trained models are already available (eg [here](https://github.com/alessiospuriomancini/cosmopower/tree/main/cosmopower/trained_models) and [here](https://github.com/cosmopower-organization)). While they may not match your needs exactly it can be a helpful starting point to build the pipeline and make the necessary modifications to MontePython before investing the time & CPU hours into training your own models.

As for the second question, no. CLASS is not compatible with autodiff and hence you will not be able to use JAX to get gradients from it. Currently, no differentiable Boltzmann solvers that calculate CMB Cls are available within python/JAX so emulators are the only solution for now I'm afraid.

**Comment by nnedelko** [comment]
Hello again! I've settled on using CosmoPower and am now working on the general calls for the gradients. I need a function in `candl_mp/__init__.py` that translates `(self, cosmo, data`) into the result of `jax.value_and_grad(self.candl_like.log_like(pars_for_like))`. Am I correctly understanding that I can just implement a `loglkl_and_grad` function that is identical to `loglkl` except for the last two lines which are changed to 
```
loglkl_vg = value_and_grad(self.candl_like.log_like)
return loglkl_vg(pars_for_like)
```

**Comment by Lbalkenhol** [comment]
Hello, I think that would work though I would propose to to define a new function in `candl_mp` that simply calls `jax.value_and_grad(self.loglkl)` in order to keep things clean. [The JAX function](https://jax.readthedocs.io/en/latest/_autosummary/jax.value_and_grad.html) should allow for the specification of which arguments to differentiate wrt to, to avoid doing so for MontePython quantities that are not of interest. If this should turn out to be too complicated, what you have proposed is perfectly fine too.
One last thing to note is that jitting as in your snippet above is ineffective as one would incur the first-time-use penalty at every evaluation as the function is being re-jitted. Instead, adding a decorator before the function definition is the way to go.

**Comment by nnedelko** [comment]
Ah, I see! I'm new to JAX, so still getting my bearings.
Thanks for the advice!
